src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -123,6 +123,77 @@ public abstract partial class Compilation
 
         internal abstract void SerializePdbEmbeddedCompilationOptions(BlobBuilder builder);
 
+        /// <summary>
+        /// This method generates a string that represents the input content to the compiler which impacts 
+        /// the output of the build. This string is effectively a content key for a <see cref="Compilation"/>
+        /// with these values that can be used to identify the outputs.
+        ///
+        /// The returned string has the following properties:
+        /// - The format is undefined. Consumers should assume the format and content can change between 
+        ///   compiler versions.
+        /// - It is designed to be human readable and diffable. This is to help developers
+        ///   understand the difference between two compilations which is impacting the deterministic 
+        ///   output
+        /// - It is *not* in a minimal form. If used as a key in say a content addressable storage consumers
+        ///   should first pass it through a strong hashing function.
+        ///
+        /// Compilations which do not use the /deterministic option can still use this API but
+        /// the results will change on every invocation.
```

Comment:
notblocking, but we should discuss this.  i think IDE would prefer we could still get good deterministic keys back in this case.  we could always attempt to workaround this by always setting the 'deterministic' option to true before calling you... but it's a bit fugly.

Reply:
Yeah I discussed this scenario with @RikkiGibson in another comment. One idea I had was adding an entry to `DeterministicKeyOptions` to suppress this behavior because for scenarios IDE is considering it is just likely resulting in needless cache misses.

Reply:
Is "the results will change on every invocation" actually the case, or would the result remain the same in most cases? 
The main exception is when `AssemblyVersionAttribute` is used in certain ways which necessarily result in a non-deterministic compilation (see [refout docs](https://github.com/dotnet/roslyn/blob/main/docs/features/refout.md#determinism).

Reply:
> Is "the results will change on every invocation" actually the case, or would the result remain the same in most cases?

There is a 100% chance the EXE / DLL will change on every invocation. That is because lacking the `/determinism` flag the MVID will be effectively `Guid.NewGuid()`.